kernel/tests/golden_tables.rs

```diff
@@ -0,0 +1,441 @@
+}
+
+// copypasta from DAT
+// some things are equivalent, but don't show up as equivalent for `==`, so we normalize here
+fn normalize_col(col: &Arc<dyn Array>) -> Arc<dyn Array> {
+    if let DataType::Timestamp(unit, Some(zone)) = col.data_type() {
+        if **zone == *"+00:00" {
+            arrow_cast::cast::cast(&col, &DataType::Timestamp(*unit, Some("UTC".into())))
+                .expect("Could not cast to UTC")
+        } else {
+            col.clone()
+        }
+    } else {
+        col.clone()
+    }
+}
+
+// copypasta from DAT
+// sort all columns in the record batch
+pub fn sort_record_batch(batch: RecordBatch) -> RecordBatch {
+    // Sort by all columns
+    let mut sort_columns = vec![];
+    for col in batch.columns() {
+        match col.data_type() {
+            DataType::Struct(_) | DataType::List(_) | DataType::Map(_, _) => {
+                // can't sort structs, lists, or maps
+            }
+            _ => sort_columns.push(SortColumn {
+                values: col.clone(),
+                options: None,
+            }),
+        }
+    }
+    let indices = lexsort_to_indices(&sort_columns, None).expect("lexsort to indices");
+    let columns = batch
+        .columns()
+        .iter()
+        .map(|c| take(c, &indices, None).unwrap())
+        .collect();
+    RecordBatch::try_new(batch.schema(), columns).expect("create record batch")
+}
+
+// do a full table scan at the latest snapshot of the table and compare with the expected data
+async fn latest_snapshot_test(
+    engine: DefaultEngine<TokioBackgroundExecutor>,
+    table: Table,
+    expected_path: Option<PathBuf>,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let snapshot = table.snapshot(&engine, None).unwrap();
+
+    let scan = snapshot.into_scan_builder().build().unwrap();
+    let scan_res = scan.execute(&engine).unwrap();
+    let batches: Vec<RecordBatch> = scan_res
+        .into_iter()
+        .map(|sr| {
+            let data = sr.raw_data.unwrap();
+            let record_batch = to_arrow(data).unwrap();
+            if let Some(mut mask) = sr.mask {
+                let extra_rows = record_batch.num_rows() - mask.len();
+                if extra_rows > 0 {
+                    // we need to extend the mask here in case it's too short
+                    mask.extend(std::iter::repeat(true).take(extra_rows));
+                }
+                filter_record_batch(&record_batch, &mask.into()).unwrap()
+            } else {
+                record_batch
+            }
+        })
+        .collect();
+
+    let expected = read_expected(&expected_path.expect("expect an expected dir"))
+        .await
+        .unwrap();
+    let schema: ArrowSchemaRef = Arc::new(scan.schema().as_ref().try_into().unwrap());
+
+    // convert the batch +00:00 to UTC
+    let result: Vec<RecordBatch> = batches
+        .iter()
+        .map(|batch| {
+            let result = batch.columns().iter().map(normalize_col).collect();
```

Comment:
we already normalize in `assert_batches_eq`

Reply:
ah nice good catch oops :)

Reply:
ah wait we have to normalize first since otherwise it will complain in `concat_batches` that the schema doesn't match...

Reply:
fixed bug in arrow_expressions.rs so no longer needed :)